### PR TITLE
Improve SSA renaming memory usage

### PR DESCRIPTION
**Pull request by mikedn**
This change mainly replaces the use of `jitstd::list` with a simple singly list based stack that uses less memory. `jitstd::list` wastes memory both because it's a doubly linked list and because the list object is quite large (head/tail pointers, size, allocators) and there are many of them (basically one for each tracked variable). It also makes memory reuse difficult - removed nodes could be reused but to do that you'd need to make a custom `jitstd` allocator that maintains a pool of nodes. Doable, but quite a bit of work due to `jitstd` using old style C++ (pre C++ 11) allocators.

Since this is likely the last large change I'm going to make to `SsaRenameState` I went an extra mile and cleaned up the whole code - comments, names etc.

This saves ~1% overall used memory, 34% `CMK_SSA` memory and ~0.1% instructions retired.

Mem diff: https://gist.github.com/mikedn/f0be27a9432e6814245d818e1a0e0b1f
PIN data: https://1drv.ms/x/s!Av4baJYSo5pjgsAnoh04vA3ZEWWD4w

Somewhat strangely, the amount of allocated memory increases a bit - 65 allocator pages. Took me a while to figure out why. Due to the fact that the slab allocator caches pages > 64k you can end up in a situation where the old code was lucky and got a larger page (e.g. 74k in one example I looked at) and after these changes it gets a normal, 64k page and then it needs another page for the rest of 10k it needs.

No x86/x64 FX diffs.

Contributes to #15108

### CODE REVIEW

**Comment by sandreenko** [comment]
Is this PR ready for review?

**Comment by mikedn** [comment]
Nah, I want to do a bit of cleanup first.

**Comment by AndyAyersMS** [comment]
@mikedn we should start clearing up your backlogged PRs.

Is this one ready for review?

**Comment by mikedn** [comment]
Yes, it's ready. Thanks!

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib

File: src/jit/ssarenamestate.cpp [hunk 1]
```diff
@@ -33,42 +33,43 @@ void SsaRenameState::EnsureStacks()
...
+//
+unsigned SsaRenameState::Top(unsigned lclNum)
 {
-    EnsureStacks();
-    DBG_SSA_JITDUMP("[SsaRenameState::CountForUse] V%02u\n", lclNum);
+    DBG_SSA_JITDUMP("[SsaRenameState::Top] V%02u\n", lclNum);
 
+    noway_assert(stacks != nullptr);
     Stack* stack = stacks[lclNum];
     noway_assert((stack != nullptr) && !stack->empty());
```
Comment (sandreenko) [inline src/jit/ssarenamestate.cpp]: Do you think that plain asserts are not enough in this case? Looks like this function is hot and can affect throughput.

**Comment by AndyAyersMS** [comment]
Might as well trigger a retest, it's been a while.

File: src/jit/ssarenamestate.cpp (hunk 1 above)
Comment (mikedn) [reply src/jit/ssarenamestate.cpp]: That does seem strange, I'll check. Usually I don't think any kind `x != nullptr` assert is very useful when `x` is immediately deferenced.

Comment (sandreenko) [reply src/jit/ssarenamestate.cpp]: > Usually I don't think any kind x != nullptr assert is very useful when x is immediately deferenced.

I was thinking about that and I am not sure, noway_assert(false) in release will be caught and the method will be recompiled with min_opts. But will VM handle a SIGSEV?

Comment (mikedn) [reply src/jit/ssarenamestate.cpp]: It's JIT's `jitNativeCode` that catches faults and drops to minopts, at least on Windows. I'd expect it also does this on Linux due to the use of `PAL_TRY`.

Comment (mikedn) [reply src/jit/ssarenamestate.cpp]: I remember now, I actually checked back then to see what happens if an access violation occurs and the result is not what I expected - the process crashes with a stack overflow exception. I've never been able to figure out, it's difficult to follow what exactly happens in debugger.

In the end I concluded that it's probably safer to just keep `noway_asserts` here. I don't really expect these asserts to ever fire, but that depends on liveness not messing things up and it's kind of difficult to be 100% confident that it won't happen.

So for the time being I think it's best to keep these. Faster code is good but safer code is even better.

**Pull request merged**